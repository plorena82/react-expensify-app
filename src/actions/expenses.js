import uuid from 'uuid';
import  database from '../firebase/firebase';

//we have multiple actions
/* 
we use Thunk because that allows use to use functions inside of our action generators.

Which in turn allows us to make asynchronous code for storing the stuff in firebase and upon success(resolve), dispatching the action to store that expense in our Redux store?

Also, because our promise was resolved, we now have access to reference of that expense and thus access to the firebase key/id and add store that as well in Redux
*/

//ADD_EXPENSE
export const addExpense = (expense) =>  ({
    type: 'ADD_EXPENSE',
    expense
});

//thunk function middleware to insert first in firebase db, then create the action addExpense to dispatch to the redux store
export const startAddExpense = (expenseData= {}) =>{

    return (dispatch,getState)=> {
        const  {
            description='',
            note='', 
            amount=0, 
            createdAt= 0
        } = expenseData;
        const uid = getState().auth.uid;
        const expense  = {description, note, amount, createdAt};
        return database.ref(`users/${uid}/expenses`).push(expense).then((snap) => {
            dispatch(addExpense({
                id: snap.key, 
                ...expense
            }));
          
        });
    };
};

// REMOVE_EXPENSE with this action we are creating the object passed to the reducer
export const removeExpense = ( {id} ={}  )  =>  ({
type: 'REMOVE_EXPENSE',
id
});

export const startRemoveExpense = ({id} ={}) =>{
    return (dispatch,getState) =>{//I could have use database.ref(`users/${uid}/expenses/${id}`).remove

         const uid= getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).set(null).then(()=>{
            dispatch(removeExpense({id}));
        });
    };
};


//EDIT_EXPENSE
export const editExpense = ( id, updates )  =>  ({
type: 'EDIT_EXPENSE',
id, 
updates    
});

export const startEditExpense = (id, updates)=>{
    return (dispatch,getState) =>{
        const uid= getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).set(updates).then(()=>{
            dispatch(editExpense(id, updates));
        });
    };
};

//SET_EXPENSES for setting into the store
export const setExpenses = (expenses) =>( {
    type:'SET_EXPENSES',
    expenses
});

// action use to fetch the expenses from firebase and performs dispatch of the SetExpenses action obj
export const startSetExpenses = () => {
     return (dispatch,getState) =>{
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot)=>{
            const expenses =[];
            snapshot.forEach((childSnap)=>{
                expenses.push({
                    id: childSnap.key,
                    ...childSnap.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
     }   

};