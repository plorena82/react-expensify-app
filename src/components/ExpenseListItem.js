
import React from 'react';
import { NavLink} from 'react-router-dom';


/* I implemented with props but it s better to use deconstruction
const ExpenseListItem = (props) => (

    <div>
        <p>{props.expense.description} { " " } {props.expense.amount} { " " } {props.expense.createdAt}    </p>         
        <button onClick={(id)=>{

            props.dispatch(removeExpense(id));
        }
        }>
            Remove
        </button> 
    </div>
);
*/

export const ExpenseListItem = ({id, description, amount, createdAt, dispatch}) => (

    <div>
        <h3><NavLink to={`/edit/${id}`}   activeClassName="is-active"> {description} </NavLink></h3>
        <p>{ " " } {amount} { " " } {createdAt}    </p>      
    </div>
);

export default ExpenseListItem;