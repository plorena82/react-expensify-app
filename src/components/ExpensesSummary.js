import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';

import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => (

    <div>
        
            { 
                
                props.expensesCount === 0 ? (
                    <p>Total: 0</p>
                ): (
                    props.expensesCount === 1 ? (
                        <p>Viewing 1 expense totalling {numeral(props.totalExpenses/100).format('$0,0.00')}</p>
                    ): (
    
                        <p>Viewing {props.expensesCount} expenses totalling {numeral(props.totalExpenses/100).format('$0,0.00')}</p>
                    )

                )

            }
          
       
      
            </div>
 ); 



const mapStateToProps = (state) =>{
    const visibleExpenses=  selectExpenses(state.expenses,state.filters)
    return{
        expensesCount:visibleExpenses.length,
        totalExpenses: selectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary); 