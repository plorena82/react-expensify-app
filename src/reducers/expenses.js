
const defaultExpenseReducerState = [];

export default (state = defaultExpenseReducerState, action) =>{
   
    switch(action.type){
        case 'ADD_EXPENSE': 
            // we can use this return state.concat(action.expense); OR 
            return [ ...state, action.expense]; //this is called ES spread operator, which doesnot change the original state array, we are returning the state array plus the new expense
        case 'REMOVE_EXPENSE': //we are using filter function of array which does not change the original state array, it returns a subarray based on condition
            return state.filter( ({id})=> id != action.id );
             //this also can be done destructuring the item of state array
             // return state.filter( ( {id} ) =>  id != action.expense.id );
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id == action.id){
                    return{
                       ...expense,
                       ...action.updates 
                    };
                }
                else{
                    console.log(action.id);
                    return expense;
                }
            });
        case 'SET_EXPENSES'://no matter the previous status, we return the expenses in action
            return action.expenses;
        default:
            return state;
    }
};