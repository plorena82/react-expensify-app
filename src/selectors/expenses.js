
import moment from 'moment';
//timestamps miliseconds since January 1st 1970  (unix epoch)

export default (expenses, {text, sortBy, startDate, endDate}) =>{
    return expenses.filter( (expense)=>{ 
            const createAtMoment = moment(expense.createdAt);
            const startDateMatch = startDate ? startDate.isSameOrBefore(createAtMoment, 'day'):true; //true means no filtering
            const endDateMatch = endDate ? endDate.isSameOrAfter(createAtMoment):true;
            //old approach //filter (returns the expenses where the startDate/end date not a number and the expenses whose createdAt date is between startdate and endDAte. IF undefined, it s like we dont have a filter )
                //  const startDateMatch = typeof startDate != 'number' || expense.createdAt >= startDate ;
            // const endDateMatch = typeof endDate != 'number' || expense.createdAt <= endDate ;
                ///
                const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

                return  startDateMatch && endDateMatch && textMatch;
    }).sort((expA, expB)=>{
        if(sortBy == 'date'){ //we want to show the recents date first a <b put b first =1, -1 a is greater than b then put a first return -1
         return expA.createdAt <  expB.createdAt ? 1:-1 ;
        }
        else if(sortBy == 'amount'){//big amounts first
            return expA.amount < expB.amount? 1: -1;
        }

    });

};