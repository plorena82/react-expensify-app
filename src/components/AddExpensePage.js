
import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense} from '../actions/expenses.js';


export class  AddExpensePage extends React.Component{
    onSubmit= (expense)=>{

        //props.dispatch(addExpense(expense)); instead of doing dispatch here, we are doing in the mapDispatchToProps,, to have a clean onSubmit call for spy functions
        this.props.startAddExpense(expense); 
        this.props.history.push('/'); // redirects to home
    };

    render(){
        return(
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">
                             Add Expense
                        </h1>
                    </div>
                </div>

                <div className="content-container">
                    <ExpenseForm 
                    onSubmit={this.onSubmit}
                    />
                </div>
             </div>
        );
    }
}
/* OLD stateless function approach.. using function, replaced by class 
const AddExpensePage = (props) => (
    <div>
        <p>Add Expense</p>
        <ExpenseForm 
        onSubmit= {(expense)=>{

            //props.dispatch(addExpense(expense)); instead of doing dispatch here, we are doing in the mapDispatchToProps,, to have a clean onSubmit call for spy functions
            props.onSubmit(expense); 
            props.history.push('/'); // redirects to home
        }}
        />
    </div>
);*/

//map the dispatch function to addExpense function attribute of a prop object
const mapDispatchToProps = (dispatch)=> ({
    
    startAddExpense: (expense)=> dispatch(startAddExpense(expense))
    
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);