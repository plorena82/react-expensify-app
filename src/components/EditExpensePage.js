

import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startEditExpense,startRemoveExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component{

    onSubmitEdit= (expense)=>{
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/'); // redirects to home 
    };

    onClickRemove = ()=>{
        this.props.startRemoveExpense({id:this.props.expense.id});
        this.props.history.push('/'); 
    };

    render(){
        return (
            <div>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmitEdit}  
                />
                <button onClick={this.onClickRemove}>
                    Remove
                </button>  
            </div>
        );
    };
}
//props.match is an object inside tthe Router which contains the parameters of the URL, inside params, we have our id
/*const EditExpensePage = (props) => {
    
    return (
        <div>
            <ExpenseForm 
                expense={props.expense}
                onSubmit={(expense) =>{
                    props.dispatch(editExpense(props.expense.id, expense));
                    props.history.push('/'); // redirects to home    
                }}  
            />
            <button onClick={()=>{
                props.dispatch(removeExpense({id : props.expense.id } ));
                props.history.push('/'); 
            }}>
                Remove
            </button>  
        </div>
    );
}*/

const mapStateToProperties = (state,props)=>{
    return {
        expense: state.expenses.find((expense) => {
            return expense.id == props.match.params.id;
        } )
    };
}
const mapDispatchToProps = (dispatch,props)=>({
    
    startEditExpense: (id,expense)=> dispatch(startEditExpense(id, expense)), 
    startRemoveExpense:  (data)=> dispatch(startRemoveExpense(data))
    
});

export default connect(mapStateToProperties,mapDispatchToProps)(EditExpensePage);