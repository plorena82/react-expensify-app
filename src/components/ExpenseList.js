import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


 export const ExpenseList = (props) => (

    <div className="content-container">
            <div className="list-header">
                <div className="show-for-mobile">Expenses</div>{/*show only Wxpenses in Mobile */}
                <div className="show-for-desktop">Expense</div>
                <div className="show-for-desktop">Amount</div>
            </div>
            <div className="list-body">
                {
                
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No expenses</span>
                    </div>
                ): (

                    props.expenses.map((expense)=> (
                        <ExpenseListItem 
                            key={expense.id}
                            {...expense}
                        />
                    ) )
                )
            
            }
            </div>            
    </div>
); 



 /* this code was inside div 
                 I implemented with props but it s better to use deconstruction 
                  props.expenses.map((expense)=> (
                <ExpenseListItem 
                    key={expense.id}
                    expense = {expense}
                />
            )


            )*/


//this HOC component function allows us to obtain from store the attributes
const mapStateToProps = (state) =>{
    return{
        expenses: selectExpenses(state.expenses,state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList); //we are indicating to connect using mapStateToProps to obtain the properties from store, and use ExpenseList as the result render component