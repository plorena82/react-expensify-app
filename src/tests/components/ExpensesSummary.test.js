import React from 'react';
import { shallow } from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';


test('should render expense summary with total 0 , no elements', ()=>{
    const wrapper = shallow(<ExpensesSummary expenses={[]} totalExpenses={0}/>);
    expect(wrapper).toMatchSnapshot();
});


test('should render expense summary with total 109.50 , 1 element', ()=>{
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} totalExpenses={10950}/>);
    expect(wrapper).toMatchSnapshot();
});


test('should render expense summary with total 156.45 , 3 elements', ()=>{
    const wrapper = shallow(<ExpensesSummary expenses={expenses} totalExpenses={15645}/>);
    expect(wrapper).toMatchSnapshot();
});