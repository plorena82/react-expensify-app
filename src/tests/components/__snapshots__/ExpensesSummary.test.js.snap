// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary with total 0 , no elements 1`] = `
<div>
  <p>
    Viewing 
     expenses totalling 
    $0.00
  </p>
</div>
`;

exports[`should render expense summary with total 109.50 , 1 element 1`] = `
<div>
  <p>
    Viewing 
     expenses totalling 
    $109.50
  </p>
</div>
`;

exports[`should render expense summary with total 156.45 , 3 elements 1`] = `
<div>
  <p>
    Viewing 
     expenses totalling 
    $156.45
  </p>
</div>
`;
